### PR TITLE
[NPU] Qwen3-Omni talker support

### PR DESCRIPTION
**Pull request by Makcum888e**
## Motivation

Support Qwen3-Omni talker on Ascend NPU to enable audio output

## Modifications

replaced `torch.cuda` with general `torch.get_device_module()`
added new functions to `current_platform`
- `get_intra_node_transport` to allow every platform choose their backend for move between devices on the same node
- `get_fused_qk_norm_rope` to import sgl-kernel only for CUDA
- `apply_model_worker_backend_policy` moved from `model_worker` as it has many CUDA related adjustment

`normalize_quantization` and `model_config_has_moe` moved from `model_worker` to `utils/misc.py`

## Accuracy Test

```
sgl-omni serve --model-path /home/weights/Qwen/Qwen3-Omni-30B-A3B-Instruct/ --thinker-cuda-graph off --thinker-gpus 2,3 --thinker-tp-size 2 --talker-gpu 1 --code2wav-gpu 0 --talker-cuda-graph off --image-encoder-gpus 0
```
```
curl -X POST http://localhost:8000/v1/chat/completions -H "Content-Type: application/json" -d '{"model": "qwen3-omni", "messages": [{"role": "user", "content": "Say the word \"generate\" exactly ten times in a row. Do not add any other words. Example: one, one, one... Repeat it exactly 10 times."}], "temperature": 0.01, "top_p": 0.1, "presence_penalty": 0.0, "frequency_penalty": 0.0, "modalities": ["text", "audio"]}' -o output.json
```

## Checklist

- [x] Format your code according with pre-commit.
- [ ] Add unit tests.
- [ ] Update documentation / docstrings / example tutorials as needed.
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the
`run-ci` label. Once labeled, every subsequent push re-triggers CI as
long as the label remains. Use `/tag-and-rerun-ci higgs` or
`/tag-and-rerun-ci moss` to select a TTS CI model, and
`/tag-and-rerun-ci fun-asr` or `/tag-and-rerun-ci qwen3-asr` to select an ASR
CI model. One selector from each family can be combined, for example
`/tag-and-rerun-ci moss fun-asr`. Draft PRs are skipped even if labeled.

### CODE REVIEW

File: sglang_omni/comm/router.py [hunk 1]
```diff
@@ -77,7 +78,7 @@ def _physical_outbound(self, target: str) -> TransportKind:
         if target in self.remote_stage_names:
             return TransportKind.MOONCAKE
         if self.self_is_gpu and target in self.gpu_stage_names:
-            return TransportKind.CUDA_IPC
+            return current_platform.get_intra_node_transport()
```
Comment (Ratish1) [inline sglang_omni/comm/router.py:84]: The default talker_ar and code2wav stages run in separate processes on the same device, so code2wav is added to _direct_cuda_ipc_targets. SGLang’s NPU bootstrap also makes an NPU tensor report is_cuda=True, which causes the direct CUDA-IPC branch in Stage._send_stream_chunk() to return before this platform transport policy is consulted.

Could we also require current_platform.get_intra_node_transport() == TransportKind.CUDA_IPC inside can_use_direct_cuda_ipc()? That keeps the existing CUDA shortcut but ensures NPU follows the SHM policy selected here.

File: sglang_omni/platforms/cuda.py [hunk 2]
```diff
@@ -43,6 +63,126 @@ def get_stage_process_env(
...
+            )
+            moe_runner_backend = server_args.moe_runner_backend
+
+        if (
+            is_qwen3_omni_arch
+            and effective_quantization == "fp8"
+            and has_moe
+            and moe_runner_backend == "auto"
+            and has_native_fp8_block_quant
+            and _is_fp8_cutlass_moe_supported()
```
Comment (Ratish1) [inline sglang_omni/platforms/cuda.py:137]: _is_fp8_cutlass_moe_supported() is not defined or imported in this module; its definition remains in model_worker.py. This branch therefore raises NameError for native block-FP8 Qwen3-Omni with  automatic MoE backend selection.

can we move the helper into this CUDA policy module, or into a dependency-neutral module, and update the tests to patch it at its new lookup location.

File: sglang_omni/platforms/cuda.py [hunk 3]
```diff
@@ -43,6 +63,126 @@ def get_stage_process_env(
...
+        server_quantization = normalize_quantization(server_args.quantization)
+        if server_quantization is not None:
+            effective_quantization = server_quantization
+
+        moe_runner_backend = server_args.moe_runner_backend
+        is_qwen3_omni_arch = model_arch_override in (
+            "Qwen3OmniTalker",
+            "Qwen3OmniThinkerForCausalLM",
+        )
+        if is_qwen3_omni_arch and server_args.ep_size != 1:
```
Comment (Ratish1) [inline sglang_omni/platforms/cuda.py]: This validation is an Omni ModelWorker invariant rather than CUDA backend policy. NPU now inherits the base policy and accepts ep_size > 1, while ModelWorker still constructs SGLModelRunner with moe_ep_size=1.

can we keep the Qwen3-Omni ep_size == 1 validation in common policy and leave only CUDA-specific backend selection in this override.

File: sglang_omni/models/qwen3_omni/components/code2wav_scheduler.py [hunk 4]
```diff
@@ -265,8 +266,7 @@ def _forward_codes(
         graph_eligible: bool = False,
     ) -> tuple[torch.Tensor, dict[str, Any]]:
         with torch.no_grad():
-            if self._device.type == "cuda":
-                torch.cuda.set_device(self._device)
+            torch.get_device_module().set_device(self._device)
```
Comment (Ratish1) [inline sglang_omni/models/qwen3_omni/components/code2wav_scheduler.py]: get_device_module() without an argument selects the process’s ambient accelerator rather than the scheduler’s configured device. On an accelerator host, a CPU Code2Wav scheduler therefore passes a CPU  device to CUDA/NPU set_device.

should we select the module from the device:

```python
torch.get_device_module(self._device).set_device(self._device)
```

File: sglang_omni/models/qwen3_omni/components/code2wav_cuda_graph.py [hunk 5]
```diff
@@ -88,48 +90,48 @@ def warmup(
...
         pool: Any,
-        stream: torch.cuda.Stream,
-    ) -> tuple[torch.cuda.CUDAGraph, torch.Tensor]:
-        current_stream = torch.cuda.current_stream(static_input.device)
+        stream: torch.Stream,
+    ) -> tuple[Any, torch.Tensor]:
+        current_stream = torch.get_device_module().current_stream(static_input.device)
         stream.wait_stream(current_stream)
-        graph = torch.cuda.CUDAGraph()
+        graph = torch.get_device_module().CUDAGraph()
```
Comment (Ratish1) [inline sglang_omni/models/qwen3_omni/components/code2wav_cuda_graph.py]: does this construct an NPU graph?. the standalone Code2Wav process resolves get_device_module() to torch.npu, which exposes NPUGraph, not CUDAGraph. The resulting AttributeError is caught during build, so the default graph-enabled Code2Wav configuration performs warmup and then silently disables every graph. The unchanged tensor.is_cuda validation would also reject native NPU inputs if construction were fixed.

should we implement NPU graphs in this PR or we keep this runner explicitly CUDA-only and disable Code2Wav graphs on NPU so it starts directly in eager mode?

File: tests/unit_test/qwen3_omni/test_fp8_backend_config.py [hunk 6]
```diff
@@ -372,14 +374,14 @@ def test_model_worker_backend_policy_precedence(
 
     if case.error_match:
         with pytest.raises(ValueError, match=case.error_match):
-            model_worker._apply_model_worker_backend_policy(
+            current_platform.apply_model_worker_backend_policy(
```
Comment (Ratish1) [inline tests/unit_test/qwen3_omni/test_fp8_backend_config.py]: current_platform is not imported, so this and the calls at lines 384, 417, and 761 raise NameError. The tests also still patch model_worker._is_fp8_cutlass_moe_supported, while production now looks up
that helper from the CUDA policy module.

lets exercise an explicit cuda.CUDAOmniPlatform() here and patch the helper in its module.

File: sglang_omni/platforms/cuda.py (hunk 2 above)
Comment (Makcum888e) [reply sglang_omni/platforms/cuda.py:137]: `_is_fp8_cutlass_moe_supported` moved to `cuda.py`

File: sglang_omni/platforms/cuda.py (hunk 3 above)
Comment (Makcum888e) [reply sglang_omni/platforms/cuda.py]: now I moved this logic to base class and call it with `super().` if this way is not suitable, I can create new function in `model_worker` for `ep_size` logic

File: sglang_omni/models/qwen3_omni/components/code2wav_scheduler.py (hunk 4 above)
Comment (Makcum888e) [reply sglang_omni/models/qwen3_omni/components/code2wav_scheduler.py]: done

File: sglang_omni/models/qwen3_omni/components/code2wav_cuda_graph.py (hunk 5 above)
Comment (Makcum888e) [reply sglang_omni/models/qwen3_omni/components/code2wav_cuda_graph.py]: removed this changes, NPUGraph will be supported in future PR

File: tests/unit_test/qwen3_omni/test_fp8_backend_config.py (hunk 6 above)
Comment (Makcum888e) [reply tests/unit_test/qwen3_omni/test_fp8_backend_config.py]: done

File: sglang_omni/comm/router.py (hunk 1 above)
Comment (Makcum888e) [reply sglang_omni/comm/router.py:84]: added this check `current_platform.get_intra_node_transport() == TransportKind.CUDA_IPC`, this seems like a reasonable check, however, this works before change and here's why:
In SGLang `transfer_to_npu` is used for replace some torch methods - it's why `is_cuda=True` in SGLang. But these workaround is not a universal solution and often creates more problems. So we not follow this way in SGL-Diffusion. I hope we can avoid using `transfer_to_npu` in SGL-Omni. So `is_cuda` will be `False` for NPU

File: sglang_omni/platforms/cuda.py (hunk 3 above)
Comment (Ratish1) [reply sglang_omni/platforms/cuda.py]: calling super() is correct. but since this restriction comes from ModelWorker always constructing SGLModelRunner with moe_ep_size=1, do you think ModelWorker would be the better owner for this check. CUDA can still call super() for the shared quantization logic.

**Review by Ratish1 — APPROVED** [review]
LGTM

File: sglang_omni/platforms/cuda.py (hunk 3 above)
Comment (Makcum888e) [reply sglang_omni/platforms/cuda.py]: moved this logic to model_worker

**Comment by Makcum888e** [comment]
Benchmark results A3 2 cards (4 devices) - 2 for thinker, 1 for talker, 1 for others:
Qwen3-Omni server command:
```
sgl-omni serve --model-path /path/to/Qwen/Qwen3-Omni-30B-A3B-Instruct/ --thinker-gpus 2,3 --thinker-tp-size 2 --talker-gpu 1 --code2wav-gpu 0 --port 8764 --image-encoder-gpus 0 --talker-cuda-graph off
```
Qwen3-ASR server command (need to disable cuda graph for now):
```
sgl-omni serve --model-path /path/to/Qwen/Qwen3-ASR-1.7B/ --port 8765 --model-name Qwen/Qwen3-ASR-1.7B
```
generate command:
```
python -m benchmarks.eval.benchmark_omni_seedtts --meta zhaochenyang20/seed-tts-eval-arrow --port 8764 --output-dir results/qwen3_omni_tts_en --lang en --max-concurrency 8 --generate-only
```
transcribe command:
```
python -m benchmarks.eval.benchmark_omni_seedtts --meta zhaochenyang20/seed-tts-eval-arrow --port 8765 --output-dir results/qwen3_omni_tts_en --lang en --max-concurrency 128 --transcribe-only --asr-model-path /path/to/Qwen/Qwen3-ASR-1.7B/
```
generate `--max-concurrency 8`:
```
============================================================
                 TTS Speed Benchmark Result                 
============================================================
  Model:                         qwen3-omni
  Dataset:                       zhaochenyang20/seed-tts-eval-arrow
  Completed requests:            1088
  Failed requests:               0
------------------------------------------------------------
  Latency mean (s):              7.079
  Latency median (s):            6.855
  Latency p95 (s):               10.463
  Latency p99 (s):               12.875
  RTF mean:                      1.9171
  RTF median:                    1.9031
  RTF p95:                       2.0946
  RTF p99:                       2.251
  Audio duration mean (s):       3.72
  Audio throughput (s/s):        4.193
  Output throughput (tok/s):     16.1
  Output tokens/request-s:       2.0
  Output tokens (mean):          14
  Output tokens (total):         15506
  Prompt tokens (mean):          31
  Prompt tokens (total):         33889
  Throughput (req/s):            1.127
============================================================

============================================================
                  TTS WER Benchmark Result                  
============================================================
  TTS model:                     qwen3-omni
  Language:                      en
  Evaluated / Total:             1088/1088
  Skipped:                       0
------------------------------------------------------------
  WER (corpus, micro-avg):       0.0224 (2.24%)
------------------------------------------------------------
  WER per-sample mean:           0.0219 (2.19%)
  WER per-sample median:         0.0000
  WER per-sample std:            0.1068
  WER per-sample p95:            0.1111
  WER per-sample max:            2.8571 (285.71%)
  WER corpus (excl >50%):        0.0181 (1.81%)
  >50% WER samples:              2 (0.2%)
------------------------------------------------------------
  Latency mean (s):              7.078811764705883
  Latency p95 (s):               10.463155
  RTF mean:                      1.9171511604860578
  Throughput (req/s):            1.127
  Audio duration mean (s):       3.7205029411764703
============================================================


============================================================
                 ASR Speed Benchmark Result                 
============================================================
  ASR model:                     /home/weights/Qwen/Qwen3-ASR-1.7B/
  Evaluated / Total:             1088/1088
  ASR concurrency:               32
  Skipped:                       0
------------------------------------------------------------
  ASR latency mean (s):          2.068171922326602
  ASR latency median (s):        2.123109660024056
  ASR latency p95 (s):           2.6636261259933236
  ASR latency p99 (s):           2.9315928031684573
  ASR RTF mean:                  0.5932765885553406
  ASR RTF median:                0.5876068490241582
  ASR total time (s):            70.76445655996213
  ASR latency sum (s):           2250.171051491343
  ASR throughput (samples/s):    15.374950263033325
  Audio processed (s):           4047.9072000000024
============================================================
```
generate `--max-concurrency 128`:
```
============================================================
                 TTS Speed Benchmark Result                 
============================================================
  Model:                         qwen3-omni
  Dataset:                       zhaochenyang20/seed-tts-eval-arrow
  Completed requests:            1088
  Failed requests:               0
------------------------------------------------------------
  Latency mean (s):              44.053
  Latency median (s):            43.854
  Latency p95 (s):               54.079
  Latency p99 (s):               59.652
  RTF mean:                      12.9288
  RTF median:                    12.0078
  RTF p95:                       20.6837
  RTF p99:                       30.5686
  Audio duration mean (s):       3.727
  Audio throughput (s/s):        10.209
  Output throughput (tok/s):     39.0
  Output tokens/request-s:       0.3
  Output tokens (mean):          14
  Output tokens (total):         15500
  Prompt tokens (mean):          31
  Prompt tokens (total):         33889
  Throughput (req/s):            2.739
============================================================

============================================================
                  TTS WER Benchmark Result                  
============================================================
  TTS model:                     qwen3-omni
  Language:                      en
  Evaluated / Total:             1088/1088
  Skipped:                       0
------------------------------------------------------------
  WER (corpus, micro-avg):       0.1581 (15.81%)
------------------------------------------------------------
  WER per-sample mean:           0.1563 (15.63%)
  WER per-sample median:         0.0000
  WER per-sample std:            0.3678
  WER per-sample p95:            1.0000
  WER per-sample max:            3.3333 (333.33%)
  WER corpus (excl >50%):        0.0252 (2.52%)
  >50% WER samples:              144 (13.2%)
------------------------------------------------------------
  Latency mean (s):              44.053469301470585
  Latency p95 (s):               54.07926499999998
  RTF mean:                      12.92879441904357
  Throughput (req/s):            2.739
  Audio duration mean (s):       3.7269735294117647
============================================================


============================================================
                 ASR Speed Benchmark Result                 
============================================================
  ASR model:                     /home/weights/Qwen/Qwen3-ASR-1.7B/
  Evaluated / Total:             1088/1088
  ASR concurrency:               32
  Skipped:                       0
------------------------------------------------------------
  ASR latency mean (s):          2.1284455432448395
  ASR latency median (s):        2.1084511050139554
  ASR latency p95 (s):           2.95846289798792
  ASR latency p99 (s):           3.4191481200890825
  ASR RTF mean:                  0.614519622788703
  ASR RTF median:                0.5830751223766153
  ASR total time (s):            72.7999532300164
  ASR latency sum (s):           2315.748751050385
  ASR throughput (samples/s):    14.945064546434391
  Audio processed (s):           4054.9471999999996
============================================================
```
